### PR TITLE
[merged] Add trust policy management workflow

### PR DESCRIPTION
**Pull request by aweiteka**
TODO:

- [x] Add tests
- [x] Configure registries.d file ~~(pending #603)~~
- [x] Add man page
- [x] Add bash completions
- [x] Add default policy control

**Example output**

```
$ cat /etc/containers/policy.json
{
    "default": [
        {
            "type": "insecureAcceptAnything"
        }
    ]
}
$ sudo atomic  trust add registry.example.com:5000 --pubkeys /etc/pki/containers/aweiteka@gmail.com
Added trust policy for atomic transport registry.example.com:5000
$ sudo atomic trust add docker.io/aweiteka --pubkeys /etc/pki/containers/aweiteka@gmail.com --sigstoretype docker
Added trust policy for docker transport docker.io/aweiteka
$ sudo atomic trust add docker.io --sigstoretype docker --type reject
Added trust policy for docker transport docker.io
$ sudo atomic trust add docker.io/aweiteka --pubkeys /etc/pki/containers/aweiteka@redhat.com --sigstoretype docker
Trust policy already defined for docker:docker.io/aweiteka
Do you want to overwrite? (y/N) y
Added trust policy for docker transport docker.io/aweiteka
$ cat /etc/containers/policy.json
{
    "default": [
        {
            "type": "insecureAcceptAnything"
        }
    ],
    "transports": {
        "docker": {
            "docker.io": [
                {
                    "type": "reject"
                }
            ],
            "docker.io/aweiteka": [
                {
                    "keyType": "GPGKeys",
                    "type": "signedBy",
                    "keyPath": "/etc/pki/containers/aweiteka@redhat.com"
                }
            ]
        },
        "atomic": {
            "registry.example.com:5000": [
                {
                    "keyType": "GPGKeys",
                    "type": "signedBy",
                    "keyPath": "/etc/pki/containers/aweiteka@gmail.com"
                }
            ]
        }
    }
}
```

### CODE REVIEW

**Comment by aweiteka** [comment]
/cc @rhatdan 
@mtrmac how is this for a workflow?

**Comment by aweiteka** [comment]
Commands wishlist:
`$ atomic trust show` <- output pretty-print policy summary
`$ atomic trust validate` <- check json compliant with skopeo / containers/image

**Comment by mtrmac** [comment]
@aweiteka Nice. Noting that this is trust only, it does not set up the sigstore URLs. It would be nice to have a single command to copy&paste; this does work like that for Atomic, not for docker+sigstore.

Possible extension: allow specifying keys as `https` (only! fetched with cert validation!) URLs, and store the fetched keys directly into the policy as `keyData`

**Comment by mtrmac** [comment]
`$ atomic trust validate` = `run any skopeo copy` right now; could be extracted into a separate utility that doesn’t require the copy.  OTOH the quality of error messages from the containers/image policy parser is probably pretty bad.

**Comment by aweiteka** [comment]
> allow specifying keys as https (only! fetched with cert validation!) URLs

Yes, there is no need to consider supporting non-TLS transport of this type of data.

> store the fetched keys directly into the policy as keyData

Cool!

**Comment by mtrmac** [comment]
> `$ atomic trust validate` = run any `skopeo copy` right now; 

To expand on this, containers/image is pretty paranoid about parsing the file, refusing any unknown keys and any duplicates, and so on. That makes it a bit easier to write editing tools because they don’t have to worry about these corner cases, OTOH I am not offhand sure how to write such a parser in Python. Of course reading such a strictly-verified file with a simple `json.load` is perfectly fine, but `json.load` accepting a file does not guarantee that containers/image will.

**Comment by aweiteka** [comment]
@baude this is complete. Please review.

File: Atomic/util.py [hunk 1]
```diff
@@ -365,7 +365,7 @@ class DockerObjectNotFound(ValueError):
     def __init__(self, msg):
         super(DockerObjectNotFound, self).__init__("Unable to associate '{}' with an image or container".format(msg))
 
-def get_atomic_config():
+def get_atomic_config(ATOMIC_CONF=ATOMIC_CONF):
```
Comment (aweiteka) [inline Atomic/util.py]: This is for overriding test config (see tests/unit/test_trust.py). Throws pylint warning. Any ideas how to do differently are welcome.

File: Atomic/trust.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+import os
+import argparse
+import json
+import yaml
+
+ATOMIC_CONFIG = util.get_atomic_config()
+
+def cli(subparser):
+    # atomic trust
+    pubkeys_dir = util.get_atomic_config_item(['pubkeys_dir'], atomic_config=ATOMIC_CONFIG)
```
Comment (rhatdan) [inline Atomic/trust.py]: No need for "atomic_config=ATOMIC_CONFIG", it is already the default.

File: Atomic/trust.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
+from . import util
+from . import Atomic
+import os
+import argparse
+import json
+import yaml
+
+ATOMIC_CONFIG = util.get_atomic_config()
```
Comment (rhatdan) [inline Atomic/trust.py]: I don't think we need this.  Already default.

File: Atomic/trust.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    removep.add_argument("--sigstoretype", dest="sigstoretype", default="docker",
+                         choices=['dir', 'docker', 'atomic'],
+                         help=sigstore_help)
+    removep.set_defaults(_class=Trust, func="remove")
+
+class Trust(Atomic):
+    def __init__(self):
+        super(Trust, self).__init__()
+        self.policy_filename = "/etc/containers/policy.json"
+        self.atomic_config = util.get_atomic_config()
```
Comment (rhatdan) [inline Atomic/trust.py]: Is this variable ever used?

File: Atomic/trust.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    commonp.add_argument("--type", dest="trust_type", default="signedBy",
+                         choices=['signedBy', 'insecureAcceptAnything', 'reject'],
+                         help="Trust type (default: signedBy)")
+    commonp.add_argument("--keytype", dest="keytype", default="GPGKeys",
+                         help="Public key type (default: GPGKeys)")
+    commonp.add_argument("--sigstore", dest="sigstore",
+                         help=_("URL and path of remote signature server, "
+                                "https://sigstore.example.com/signatures. "
+                                "Ignored with 'atomic' sigstoretype."))
+    commonp.add_argument("registry",
```
Comment (rhatdan) [inline Atomic/trust.py]: Do you want to add any short options?  -s, -t, -k?

File: Atomic/trust.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                         help=_("URL and path of remote signature server, "
+                                "https://sigstore.example.com/signatures. "
+                                "Ignored with 'atomic' sigstoretype."))
+    commonp.add_argument("registry", 
+                         help=registry_help)
+    subparsers = trustp.add_subparsers()
+    addp = subparsers.add_parser("add", parents=[commonp],
+                                 help="Add a new trust policy for a registry")
+    addp.set_defaults(_class=Trust, func="add")
+    removep = subparsers.add_parser("remove",
```
Comment (rhatdan) [inline Atomic/trust.py]: We use "delete" for atomic images and atomic containers, should we use it for trust?

File: Atomic/trust.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+                         help=registry_help)
+    removep.add_argument("--sigstoretype", dest="sigstoretype", default="docker",
+                         choices=['dir', 'docker', 'atomic'],
+                         help=sigstore_help)
+    removep.set_defaults(_class=Trust, func="remove")
+
+class Trust(Atomic):
+    def __init__(self):
+        super(Trust, self).__init__()
+        self.policy_filename = "/etc/containers/policy.json"
```
Comment (rhatdan) [inline Atomic/trust.py]: How about 
```
def __init__(self, policy_filename="/etc/containers/policy.json"
        super(Trust, self).__init__()
        self.policy_filename = policy_filename
````

Then your tests below can just call 

trust=Trust("trust.con")

File: Atomic/trust.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+                confirm = None
+                if self.args.assumeyes:
+                    confirm = "yes"
+                else:
+                    confirm = util.input("Trust policy already defined for %s:%s\nDo you want to overwrite? (y/N) " % (self.args.sigstoretype, self.args.registry))
+                if not "y" in confirm.lower():
+                    util.write_out("Trust policy not modified")
+                    exit(0)
+            payload = []
+            if self.args.pubkeys:
```
Comment (rhatdan) [inline Atomic/trust.py]: Isn't pubkeys a list?  If not it should be.  Not a space separated blob.

File: Atomic/util.py (hunk 1 above)
Comment (rhatdan) [inline Atomic/util.py]: atomic_conf=None

if not atomic_conf:

**Comment by rhatdan** [comment]
Need bash completions.

File: Atomic/trust.py (hunk 6 above)
Comment (aweiteka) [inline Atomic/trust.py]: I like "remove" but it's probably best to be consistent with the other args in the tool.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Comment by rhatdan** [comment]
@aweiteka your tests are failing.

**Comment by baude** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**